File: timeentryservice/src/main/java/org/timeasy/models/TimeEntry.java
package org.timeasy.models;

import org.hibernate.annotations.Type;

import javax.persistence.Entity;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;

import java.time.Instant;
import java.util.UUID;

@Entity
public class TimeEntry {

    @Id
    @Type(type = "pg-uuid")
    private UUID id;
    private String description;
    private Instant startTime;
    private Instant endTime;
    @ManyToOne
    @JoinColumn(name = "project_id", columnDefinition = "UUID")
    private Project project;
    private String userId;
    private Instant createdTimeStamp = Instant.now();
    private Instant updatedTimeStamp = Instant.now();
    private boolean deleted;

    public TimeEntry() {
        id = UUID.randomUUID();
    }

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public Instant getStartTime() {
        return startTime;
    }

    public void setStartTime(Instant startTime) {
        this.startTime = startTime;
    }

    public Instant getEndTime() {
        return endTime;
    }

    public void setEndTime(Instant endTime) {
        this.endTime = endTime;
    }

    public Project getProject() {
        return project;
    }

    public void setProject(Project project) {
        this.project = project;
    }

    public String getUserId() {
        return userId;
    }

    public void setUserId(String userId) {
        this.userId = userId;
    }

    public Instant getCreatedTimeStamp() {
        return createdTimeStamp;
    }

    public void setCreatedTimeStamp(Instant createdTimeStamp) {
        this.createdTimeStamp = createdTimeStamp;
    }

    public Instant getUpdatedTimeStamp() {
        return updatedTimeStamp;
    }

    public void setUpdatedTimeStamp(Instant updatedTimeStamp) {
        this.updatedTimeStamp = updatedTimeStamp;
    }

    public boolean isDeleted() {
        return deleted;
    }

    public void setDeleted(boolean deleted) {
        this.deleted = deleted;
    }
}

File: src/pkg/domain/model/timeentry.go
package model

import (
	"time"

	"github.com/gofrs/uuid"
	"gorm.io/gorm"
)

type TimeEntry struct {
	gorm.Model
	ID          uuid.UUID `gorm:"type:uuid;primaryKey;"`
	UserId      uuid.UUID `gorm:"type:uuid;"`
	ProjectId   uuid.UUID `gorm:"type:uuid;"`
	StartTime   time.Time `gorm:"type:timestamp;"`
	EndTime     time.Time `gorm:"type:timestamp;"`
	Description string
}

func (timeEntry *TimeEntry) BeforeCreate(db *gorm.DB) error {
	id, err := uuid.NewV4()
	if err != nil {
		return err
	}
	timeEntry.ID = id
	if timeEntry.StartTime.IsZero() {
		timeEntry.StartTime = time.Now().UTC()
	}
	return nil
}
